test/helpers/position_assertions.cc

```diff
@@ -1130,102 +1130,271 @@ void ApplyCodeActionAssertion::check(const UnorderedMap<std::string, std::shared
+                               query, optionPair, assertionContents);
+        }
+        remainingOptions = matches[6].str();
     }
+    return make_shared<SymbolSearchAssertion>(filename, range, assertionLine, query, name, container, rank, uri);
+}
 
-    ADD_FAILURE_AT(string(filename).c_str(), assertionLine + 1) << fmt::format(
-        "Improperly formatted assertion. Expected 'symbol-search: \"<query>\"'. Found '{}'", assertionContents);
+bool SymbolSearchAssertion::matches(std::string_view uriPrefix, const SymbolInformation &symbol) const {
+    auto assertionLocation = getLocation(uriPrefix);
+    auto &symbolLocation = symbol.location;
+    if (uri.has_value()) {
+        if (symbolLocation->uri.find(*uri) == string::npos) {
+            return false;
+        }
+    } else {
+        if (assertionLocation->uri != symbolLocation->uri) {
+            return false;
+        }
+        if (assertionLocation->range->start->line != symbolLocation->range->start->line) {
+            return false;
+        }
+    }
+    if (name.has_value() && *name != symbol.name) {
+        return false;
+    }
+    if (container.has_value() && *container != symbol.containerName) {
+        return false;
+    }
+    return true;
+}
+namespace {
+string pluralized_count(string_view word, int count) {
+    return fmt::format("{} {}{}", count, word, (count == 1) ? "" : "s");
+}
 
-    return nullptr;
+void addFailureAtLocationWithSource(const Location &location,
+                                    const UnorderedMap<string, shared_ptr<core::File>> &sourceFileContents,
+                                    string_view uriPrefix, std::string_view message) {
+    if (!isUriATestFile(uriPrefix, location.uri)) {
+        ADD_FAILURE_AT(location.uri.c_str(), location.range->start->line + 1)
+            << fmt::format("{}\n(source unavailable for {})\n", message, location.uri);
+        return;
+    }
+    auto sourceLine = getLine(sourceFileContents, uriPrefix, location);
+    ADD_FAILURE_AT(location.uri.c_str(), location.range->start->line + 1)
+        << fmt::format("{}\n{}\n", message, prettyPrintRangeComment(sourceLine, *location.range, ""));
 }
 
-void SymbolSearchAssertion::checkAllForQuery(std::string query,
-                                             const vector<shared_ptr<SymbolSearchAssertion>> &assertions,
-                                             const UnorderedMap<string, shared_ptr<core::File>> &sourceFileContents,
-                                             LSPWrapper &lspWrapper, int &nextId, string_view uriPrefix,
-                                             string errorPrefix) {
-    const int id = nextId++;
-    const int lineWidth = 6; // ensure sort order == numerical order (if file is less than a million lines)
+void matchSymbolsToAssertions(string_view query, const vector<shared_ptr<SymbolSearchAssertion>> &assertions,
+                              vector<pair<SymbolInformation &, SymbolSearchAssertion *>> &symbolMatchStates,
```

Comment:
What would you think about managing these by data structures by copying things around? There's only ever going to be a handful of them, and seeing things like raw pointers / references / reference_wrapper / `&*assertion` etc. seems like we're over-engineering this a bit. Thoughts?

Reply:
That was what I tried at first, but these don't have copy constructors, so it seemed less intrusive for me to use references/pointers than it was to add copy constructors.  (If you'd prefer that I add copy constructors, let me know which pointers/references you'd like me to switch...)